src/engine/GroupBy.h

```diff
@@ -244,13 +264,56 @@ class GroupBy : public Operation {
   std::optional<HashMapOptimizationData> checkIfHashMapOptimizationPossible(
       std::vector<Aggregate>& aggregates);
 
+  // Extract values from `expressionResult` and store them in the rows of
+  // `resultTable` specified by the indices in `evaluationContext`, in column
+  // `outCol`.
+  static void extractValues(
+      sparqlExpression::ExpressionResult&& expressionResult,
+      sparqlExpression::EvaluationContext& evaluationContext,
+      IdTable* resultTable, LocalVocab* localVocab, size_t outCol);
+
+  // In case of aliases that contain an aggregate at the top-level of the
+  // expression tree, e.g. "AVG(?y) as ?avg", we can directly store the values
+  // calculated in our HashMap in the result table.
+  template <size_t numAggregates>
+  sparqlExpression::VectorWithMemoryLimit<ValueId> extractValuesDirectlyFromMap(
+      sparqlExpression::EvaluationContext& evaluationContext,
+      const ad_utility::HashMapWithMemoryLimit<KeyType,
+                                               ValueType<numAggregates>>& map,
+      IdTable* resultTable, size_t hashMapIndex, size_t outCol);
+
+  // Substitute the results for all aggregates in `info`. The values of the
+  // grouped variable should be at column 0 in `groupValues`.
+  template <size_t numAggregates>
+  void substituteAllAggregates(
+      std::vector<HashMapAggregateInformation>& info,
+      sparqlExpression::EvaluationContext& evaluationContext,
+      const ad_utility::HashMapWithMemoryLimit<KeyType,
+                                               ValueType<numAggregates>>& map,
+      IdTable* groupValues);
+
+  // Check if an expression is of a certain type.
+  template <class T>
+  static bool hasType(sparqlExpression::SparqlExpression* expr);
+
+  // Check if an expression is a currently supported aggregate.
+  // TODO<kcaliban> As soon as all aggregates are supported, implement and use a
+  //                `isAggregate` function in SparqlExpressions instead.
+  static bool isSupportedAggregate(sparqlExpression::SparqlExpression* expr);
+
+  // Find all aggregates for expression `expr`. Return `std::nullopt`
+  // if an unsupported aggregate is found.
+  // TODO<kcaliban> Remove std::optional as soon as all aggregates are supported
+  static std::optional<std::vector<HashMapAggregateInformation>> findAggregates(
+      sparqlExpression::SparqlExpression* expr);
+
   // Find all aggregates of expression `expr`, collecting this information in
   // the vector `info`. Returns false in case an unsupported aggregate is
   // found.
-  bool findAggregateMultiple(sparqlExpression::SparqlExpression* parent,
-                             sparqlExpression::SparqlExpression* expr,
-                             std::optional<size_t> index,
-                             std::vector<HashMapAggregateInformation>& info);
+  static bool findAggregatesImpl(
```

Comment:
The comment here is redundant.
Make it "// The recursive implementation of `findAggregates` (see above)"